pkg/apis/ceph.rook.io/v1/spec_test.go

```diff
@@ -52,7 +53,8 @@ storage:
     directories:
     - path: "/rook/dir1"
   - name: "node2"
-    deviceFilter: "^foo*"`)
+    deviceFilter: "^foo*"
+    devLinksFilter: "^/dev/disk/by-id/.*foo.*"`)
```

Comment:
Is there a technical reason why the links filter needs to be separate from device filter? From a user perspective, I think I would prefer to use `deviceFilter` for any of `sd[b-z]`, `/dev/sd[b-z]`, or `/dev/disk/by-id/.*` strings.

Reply:
I agree that extending `deviceFilter` is preferable to introducing a new filter.
I added a new filter to avoid incompatibility.
If `deviceFilter` is extended to be compared with long device names while a user has specified `deviceFilter: sd[b-z]` without caring about long names, the extended filter can mistakenly match long names containing `sde` substring, e.g. `/dev/disk/by-id/dm-name-misdefined-lv`.

https://github.com/rook/rook/pull/2845 applies the strategy of "if a path specification begins with `/dev`, compare it to the full paths of devices."
I once examined this strategy and thought that it was not applicable to device filters.
A device filter is given as a regular expression, so it may not begin with `/dev` nor `^/dev` even when it is expected to match a full path.

Now I've got an idea of "if a device filter contains `/`, compare it to full paths."

Do you have any good solutions?

Reply:
Those are good and well-defined thoughts @morimoto-cybozu. Thank you. My main goal is to choose whatever will be less confusing for users (while still considering development and maintenance effort).

With `deviceFilter` we support filtering disks where we assume the `/dev` path is present in front. e.g., `sd[b-z]` matches `/dev/sd[b-z]`. This is pretty basic and not really sufficient for the real world.

What we want from our filtering is to be able to additionally support matching disks in `/dev/disk/...`. Do we also want to support matching by `/dev/mapper`?

Is it safe to make the assumption that disks will always be in `/dev`?

-----

> Now I've got an idea of "if a device filter contains /, compare it to full paths.

I wonder what it looks like to use this from a user perspective. I think there are a few cases.
Cases:
1. Old/existing way (matching 'basic' disks)
    - `sd[b-z]`should match `/dev/sdb` through `/dev/sdz`
    - `sg[a-z][a-z]?` should match `/dev/sga` through `/dev/sgzz`
2. New way, but users want to match basic disks
    - `/dev/sd[b-z]` (match same as above)
    - `/dev/sg[a-z][a-z]?` (match same as above)
3. New way, with non-basic disks
    - `/dev/disk/by-id/.*foo.*` should match all disks that have id containing 'foo'
4. Potentially aberrant case of **3** assuming `/dev` is prepended; a user might assume that `/dev` is optional for disk matching based on the behavior defined in **1** above
    - `disk/by-id/.*foo.*` should match all disks that have id containing 'foo' still
5. New way, matching mapper disks
    - `/dev/mapper/.*foo.*` should match all mapper disk with 'foo'
6. Potentially aberrant case of **5** assuming `/dev` is prepended
    - `mapper/.*foo.*` should match all mapper disks with 'foo'
7. Try to break the logic
    - `disk` should not match anything
    - `mapper` should not match anything
    - `/dev/disk` should not match anything????
    - `/dev/mapper` should not match anything????

-----

I also want to consider what exists currently as actual behavior, not ideal behavior. Currently, by looking in the code, devices are listed with `lsblk -all`, and then they are selected for use from among this list by a pretty simple `regexp.Match`. On my local system, this reports disks only as `sd[a-z]`. I'm not sure if this would report `sg` disks at all even, but I can imagine hoping to match disks by `/dev/sg` somehow. So the current filter really is just a matcher do disks from `lsblk`'s output.

This also brings to my mind, for matching disk by path (I feel like we are sort of considering matching disks by path here?), we need to also make sure they are disks and not partitions or other unsupported types. I'm not sure how to validate that a given is a disk off the top of my head.

This becomes easier if we know we are only looking at disks in `/dev/disk/...`. If we have to support any path under `/dev/...`, things start getting harder I think.

-----

Sorry my thoughts are kind of scattered. I'm hoping to tease out both ideal behavior as well as ideal user interface here. Personally, I think using `/` as a marker for path matching might work well, but there could be breaking examples I didn't think of. We do need to still make sure we are matching the paths matched back to disks which exist and are actual `disk` types (e.g., not `part`, `crypt`, etc.) so that we aren't matching things Rook can't support. We can probably cross-reference the `lsblk` output for that. I may be repeating a lot of information you've already gotten down.

Reply:
Sorry for the late reply.

Your several cases are thought-provoking for me.  Though I proposed the idea of using "/" to switch the path matching mode, it seems confusing to users.  The current `deviceFilter` implies a prefix of "/dev" as you mentioned.  The combination of the implicit "/dev" and the matching mode switch may cause users to write aberrant and unreadable filter patterns, as you pointed in examples.

So I re-propose to add a new filter instead of extending `deviceFilter`.  This keeps `deviceFilter` simple and gives clear definition to the new filter.

By the way, `"/dev/sd[b-z]"` in your case 2 does not match "/dev/sdb" in the current `devLinksFilter` implementation.  This is because I coded to compare the filter only with `DEVLINKS` paths in the `udevadm` output.  I introduced `devLinksFilter` to use predictable and topology-aware device names, so "/dev/sdb" was out of the scope.  This is simple in coding but complicated for users with the benefit of hindsight.  I will improve the matching to cover the case 2, probably with changing the filter name to more appropriate one.

> Currently, by looking in the code, devices are listed with `lsblk -all`
(snip)
> I'm not sure if this would report `sg` disks at all even, but I can imagine hoping to match disks by `/dev/sg` somehow.
(snip)
> If we have to support any path under `/dev/...`, things start getting harder I think.

I think that the use of `lsblk` is reasonable to get the list of devices for Rook/Ceph.  `lsblk` lists all block devices on the system, and Rook/Ceph uses block devices for OSDs.  Using `KNAME`s and `DEVLINKS` names returned by `udevadm` would be sufficient for Rook/Ceph.

`/dev/sg*` are character special files to send SCSI commands to devices, e.g. in burning a CD-R, and not used by Rook/Ceph.

> We do need to still make sure we are matching the paths matched back to disks which exist and are actual disk types (e.g., not part, crypt, etc.)

The types of disks are checked in [`PopulateDeviceInfo()`](https://github.com/rook/rook/blob/v1.1.7/pkg/clusterd/disk.go#L76), which is indirectly called from [the OSD Provisioning Pod](https://github.com/rook/rook/blob/v1.1.7/pkg/daemon/ceph/osd/daemon.go#L197) to give device candidates to `getAvailableDevices()`.  My PR uses these candidates in matching as well as the current `deviceFilter` does, so the device type constraint will be kept.

Note that this check allows `crypt` and `part` even in the original version of Rook.  I think these types are valid for Rook.  In fact my use case is to construct a Rook/Ceph cluster with pre-encrypted disks.

Reply:
@BlaineEXE Any more thoughts on keeping the separate `devLinksFilter` property? Or shall we go with this?